mgmtsystem/openstack.py

```diff
@@ -474,40 +462,112 @@ def deploy_template(self, template, *args, **kwargs):
+            instance_name = instance_or_name
+            instance = self._find_instance_by_name(instance_name)
+
+        current_ip = self.current_ip_address(instance_name)
+        if current_ip is not None:
+            return current_ip
+
+        # Why while? Well, this code can cause one peculiarity. Race condition can "steal" a FIP
+        # so this will loop until it really get the address. A small timeout is added to ensure
+        # the instance really got that address and other process did not steal it.
+        # TODO: Introduce neutron client and its create+assign?
+        while self.current_ip_address(instance_name) is None:
+            free_ips = self.free_fips(floating_ip_pool)
             # We maintain 1 floating IP as a protection against race condition
             # I know it is bad practice, but I did not figure out how to prevent the race
             # condition by openstack saying "Hey, this IP is already assigned somewhere"
             if len(free_ips) > 1:
                 # There are 2 and more ips, so we will take the first one (eldest)
                 ip = free_ips[0]
-                self.logger.info("Reusing {} from pool {}".format(ip.ip, pool))
+                self.logger.info("Reusing {} from pool {}".format(ip.ip, floating_ip_pool))
             else:
                 # There is one or none, so create one.
                 try:
-                    ip = self.api.floating_ips.create(pool)
+                    ip = self.api.floating_ips.create(floating_ip_pool)
                 except (os_exceptions.ClientException, os_exceptions.OverLimit) as e:
                     self.logger.error("Probably no more FIP slots available: {}".format(str(e)))
-                    free_ips = self.free_fips(pool)
+                    free_ips = self.free_fips(floating_ip_pool)
                     # So, try picking one from the list (there still might be one)
                     if free_ips:
                         # There is something free. Slight risk of race condition
                         ip = free_ips[0]
                         self.logger.info(
                             "Reused {} from pool {} because no more free spaces for new ips"
-                            .format(ip.ip, pool))
+                            .format(ip.ip, floating_ip_pool))
                     else:
                         # Nothing can be done
                         raise NoMoreFloatingIPs("Provider {} ran out of FIPs".format(self.auth_url))
-                self.logger.info("Created {} in pool {}".format(ip.ip, pool))
+                self.logger.info("Created {} in pool {}".format(ip.ip, floating_ip_pool))
             instance.add_floating_ip(ip)
-            self.logger.info("Instance {} got a floating IP {}".format(kwargs['vm_name'], ip.ip))
 
-        if power_on:
-            self.start_vm(kwargs['vm_name'])
+            # Now the grace period in which a FIP theft could happen
+            time.sleep(safety_timer)
```

Comment:
In 99% of cases when it works properly, it would still wait the maximum time. This timeout is supposed to handle the timeframe in which the FIP picked by this loop iteration could be also picked by another process and stolen.

The better solution would be using the neutron client and its create+assign in one step. I did not want to change it though because this one is already working and I just added one more layer to protect the IP assignment :)